### PR TITLE
HDDS-10460. Refine audit logging for bucket property updation operations

### PR DESCRIPTION
**Pull request by tanvipenumudy**
## What changes were proposed in this pull request?

We should refine audit logging for operations modifying bucket properties.

**How can this be useful?**

- Critical for consumers on earlier versions of Ozone who could potentially run into known bugs: [HDDS-7449](https://issues.apache.org/jira/browse/HDDS-7449) and [HDDS-7526](https://issues.apache.org/jira/browse/HDDS-7526).
- Losing bucket replication properties/bucket encryption properties when one (re)sets quota/bucket replication configurations poses significant risks.
- It is difficult for diagnosing the root cause when one runs into such issues just by looking at the audit logs.
- Currently, the audit logs do not provide much insight into what properties have been modified while performing bucket config re(set) operations.

As of today, we are only capturing basic information such as volume, bucket, gdprEnabled, isVersionEnabled, storageType and owner properties for any given bucket.

We should also be capturing bucket quota, encryption and replication-related properties.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10460

## How was this patch tested?

- Green Git CI: https://github.com/tanvipenumudy/ozone/actions/runs/8152588034/
- The patch has been tested manually on a docker-compose cluster.

Create a bucket with replication-type: `RATIS`, replication-factor: `THREE`:
```
ozone sh bucket create voltest/bucktest --type=RATIS --replication=THREE
```
Set `quotaInBytes`, `quotaInNamespace` for the created bucket:
```
ozone sh bucket setquota --quota=1GB --namespace-quota=5 voltest/bucktest
```
`om-audit.log`:
```
2024-03-05 12:35:15,917 | INFO  | OMAudit | user=hadoop | ip=172.19.0.9 | op=UPDATE_BUCKET {volume=voltest, bucket=bucktest, gdprEnabled=null, isVersionEnabled=null, quotaInBytes=1073741824, quotaInNamespace=5} | ret=SUCCESS |
```
Set replication-config from replication-factor: `THREE` to replication-factor: `ONE`:
```
ozone sh bucket set-replication-config --type=RATIS --replication=ONE voltest/bucktest
```
`om-audit.log`:
```
2024-03-05 12:35:59,013 | INFO  | OMAudit | user=hadoop | ip=172.19.0.9 | op=UPDATE_BUCKET {volume=voltest, bucket=bucktest, gdprEnabled=null, isVersionEnabled=null, replicationType=RATIS, replicationFactor=ONE} | ret=SUCCESS |
```

### CODE REVIEW

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmBucketArgs.java [hunk 1]
```diff
@@ -187,6 +187,27 @@ public Map<String, String> toAuditMap() {
...
+          String.valueOf(this.quotaInNamespace));
+    }
+    if (this.bekInfo != null) {
+      auditMap.put(OzoneConsts.BUCKET_ENCRYPTION_KEY_NAME,
+          this.bekInfo.getKeyName());
+    }
+    if (this.defaultReplicationConfig != null) {
+      auditMap.put(OzoneConsts.REPLICATION_TYPE, String.valueOf(
+          this.defaultReplicationConfig.getType()));
+      auditMap.put(OzoneConsts.REPLICATION_FACTOR,
```
Comment (adoroszlai) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmBucketArgs.java:207]: "Factor" does not apply to EC, so I suggest using `REPLICATION_CONFIG` instead.

```suggestion
      auditMap.put(OzoneConsts.REPLICATION_CONFIG,
```

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmBucketArgs.java [hunk 2]
```diff
@@ -187,6 +187,27 @@ public Map<String, String> toAuditMap() {
     if (this.ownerName != null) {
       auditMap.put(OzoneConsts.OWNER, this.ownerName);
     }
+    if (this.quotaInBytesSet && quotaInBytes > 0 ||
+        !(this.quotaInBytes == OzoneConsts.QUOTA_RESET)) {
```
Comment (adoroszlai) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmBucketArgs.java:191]: nit: can be simplified?

```suggestion
        (this.quotaInBytes != OzoneConsts.QUOTA_RESET)) {
```

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmBucketArgs.java [hunk 3]
```diff
@@ -187,6 +187,27 @@ public Map<String, String> toAuditMap() {
...
+      auditMap.put(OzoneConsts.QUOTA_IN_BYTES,
+          String.valueOf(this.quotaInBytes));
+    }
+    if (this.quotaInNamespaceSet && quotaInNamespace > 0 ||
+        !(this.quotaInNamespace == OzoneConsts.QUOTA_RESET)) {
```
Comment (adoroszlai) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmBucketArgs.java:196]: nit: similarly

```suggestion
        (this.quotaInNamespace != OzoneConsts.QUOTA_RESET)) {
```

**Review by adoroszlai — COMMENTED** [review]
Thanks @tanvipenumudy for the patch.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java [hunk 4]
```diff
@@ -294,6 +294,7 @@ private OzoneConsts() {
   public static final String USED_NAMESPACE = "usedNamespace";
   public static final String QUOTA_IN_BYTES = "quotaInBytes";
   public static final String QUOTA_IN_NAMESPACE = "quotaInNamespace";
+  public static final String BUCKET_ENCRYPTION_KEY_NAME = "bekName";
```
[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java:297]:
```suggestion
  public static final String BUCKET_ENCRYPTION_KEY_NAME = "bucketEncryptionKey";
```

[the remark this answers is not in the document]
Comment (tanvipenumudy) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java]: Thank you for the review and suggested change!

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java [hunk 5]
```diff
@@ -294,6 +294,7 @@ private OzoneConsts() {
   public static final String USED_NAMESPACE = "usedNamespace";
   public static final String QUOTA_IN_BYTES = "quotaInBytes";
   public static final String QUOTA_IN_NAMESPACE = "quotaInNamespace";
+  public static final String BUCKET_ENCRYPTION_KEY_NAME = "bucketEncryptionKey";
```
Comment (adoroszlai) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java:297]: It turns out we can reuse existing constant:

https://github.com/apache/ozone/blob/a66ad7cfc027f70dee51692050e73cb16bc960ed/hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java#L339

**Review by adoroszlai — COMMENTED** [review]
Thanks again @tanvipenumudy for the patch.

On another look, similar change is needed in `OmBucketInfo`, which is used as source of audit data during bucket creation:

https://github.com/apache/ozone/blob/4243721dda6919e7614f8d1258027e2aec55bc28/hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmBucketInfo.java#L306-L333

Quotas and default replication config need to be added (please double-check I did not miss anything).

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java (hunk 5 above)
Comment (tanvipenumudy) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConsts.java]: Thank you @adoroszlai for pointing this out! I have reused the existing constant!

**Comment by tanvipenumudy** [comment]
> On another look, similar change is needed in OmBucketInfo, which is used as source of audit data during bucket creation
> Quotas and default replication config need to be added

Thank you @adoroszlai and @myskov for the reviews. I have filed a ticket: [HDDS-10475](https://issues.apache.org/jira/browse/HDDS-10475) for tracking the [suggested change](https://github.com/apache/ozone/pull/6329#pullrequestreview-1918970353)!

**Comment by adoroszlai** [comment]
Thanks @tanvipenumudy for the patch, @myskov for the review.

**Pull request merged**